File: source/extensions/omni.isaac.wheeled_robots/omni/isaac/wheeled_robots/nodes/OgnDifferentialController.py
# Copyright (c) 2022-2023, NVIDIA CORPORATION. All rights reserved.
#
# NVIDIA CORPORATION and its licensors retain all intellectual property
# and proprietary rights in and to this software, related documentation
# and any modifications thereto. Any use, reproduction, disclosure or
# distribution of this software and related documentation without an express
# license agreement from NVIDIA CORPORATION is strictly prohibited.
#


import numpy as np
from omni.isaac.core.utils.types import ArticulationAction
from omni.isaac.core_nodes import BaseResetNode
from omni.isaac.wheeled_robots.controllers.differential_controller import DifferentialController
from omni.isaac.wheeled_robots.ogn.OgnDifferentialControllerDatabase import OgnDifferentialControllerDatabase


class OgnDifferentialControllerInternalState(BaseResetNode):
    def __init__(self):
        self.wheel_radius = (float,)
        self.wheel_distance = (float,)
        self.controller_handle = None
        self.max_linear_speed = 1.0e20
        self.max_angular_speed = 1.0e20
        self.max_wheel_speed = 1.0e20
        self.store_joint_actions = None
        self.outputs = None
        super().__init__(initialize=False)

    def initialize_controller(self) -> None:
        self.controller_handle = DifferentialController(
            name="differential_controller",
            wheel_radius=self.wheel_radius,
            wheel_base=self.wheel_distance,
            max_linear_speed=self.max_linear_speed,
            max_angular_speed=self.max_angular_speed,
            max_wheel_speed=self.max_wheel_speed,
        )
        self.initialized = True

    def forward(self, command: np.ndarray) -> ArticulationAction:
        return self.controller_handle.forward(command)

    def custom_reset(self):
        if self.initialized:
            if self.store_joint_actions.joint_positions is not None:
                self.outputs.positionCommand = [0.0, 0.0]
            if self.store_joint_actions.joint_velocities is not None:
                self.outputs.velocityCommand = [0.0, 0.0]
            if self.store_joint_actions.joint_efforts is not None:
                self.outputs.effortCommand = [0.0, 0.0]


class OgnDifferentialController:
    """
    nodes for moving an articulated robot with joint commands
    """

    @staticmethod
    def initialize(graph_context, node):
        # Store db.outputs in a private variable of State class so we can modify the output on simulation Stop
        db = OgnDifferentialControllerDatabase(node)
        state = OgnDifferentialControllerDatabase.per_node_internal_state(node)
        state.outputs = db.outputs

    @staticmethod
    def internal_state():
        return OgnDifferentialControllerInternalState()

    @staticmethod
    def compute(db) -> bool:
        state = db.internal_state

        try:
            if not state.initialized:

                if db.inputs.wheelRadius <= 0 or db.inputs.wheelDistance <= 0:
                    db.log_warning("invalid wheel radius and distance")
                    return False
                else:
                    state.wheel_radius = db.inputs.wheelRadius
                    state.wheel_distance = db.inputs.wheelDistance

                if db.inputs.maxLinearSpeed > 0:
                    state.max_linear_speed = db.inputs.maxLinearSpeed
                elif db.inputs.maxLinearSpeed == 0:
                    state.max_linear_speed = 1.0e20

                if db.inputs.maxAngularSpeed > 0:
                    state.max_angular_speed = db.inputs.maxAngularSpeed
                elif db.inputs.maxAngularSpeed == 0:
                    state.max_angular_speed = 1.0e20

                if db.inputs.maxWheelSpeed > 0:
                    state.max_wheel_speed = db.inputs.maxWheelSpeed
                elif db.inputs.maxWheelSpeed == 0:
                    state.max_wheel_speed = 1.0e20

                state.initialize_controller()

            joint_actions = state.forward(np.array([db.inputs.linearVelocity, db.inputs.angularVelocity]))
            if joint_actions.joint_positions is not None:
                db.outputs.positionCommand = joint_actions.joint_positions
            if joint_actions.joint_velocities is not None:
                db.outputs.velocityCommand = joint_actions.joint_velocities
            if joint_actions.joint_efforts is not None:
                db.outputs.effortCommand = joint_actions.joint_efforts

            state.store_joint_actions = joint_actions

        except Exception as error:
            db.log_error(str(error))
            return False

        return True

    @staticmethod
    def release(node):
        try:
            state = OgnDifferentialControllerDatabase.per_node_internal_state(node)
        except Exception:
            state = None
            pass

        if state is not None:
            state.reset()

File: source/extensions/omni.isaac.wheeled_robots/nodes/OgnDifferentialController.cpp
// Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
//
// NVIDIA CORPORATION and its licensors retain all intellectual property
// and proprietary rights in and to this software, related documentation
// and any modifications thereto. Any use, reproduction, disclosure or
// distribution of this software and related documentation without an express
// license agreement from NVIDIA CORPORATION is strictly prohibited.
//

// clang-format off
#include "UsdPCH.h"
// clang-format on

#include <omni/isaac/utils/BaseResetNode.h>

#include <OgnDifferentialControllerDatabase.h>
#include <cmath>

namespace omni
{
namespace isaac
{
namespace wheeled_robot
{

class OgnDifferentialController : public BaseResetNode
{

public:
    static bool compute(OgnDifferentialControllerDatabase& db)
    {
        auto& state = db.perInstanceState<OgnDifferentialController>();
        state.nodeObj = db.abi_node();

        if (!state.mInitialized)
        {
            if (db.inputs.wheelRadius() <= 0 || db.inputs.wheelDistance() <= 0)
            {
                db.logWarning("invalid wheel radius and distance");
                return false;
            }
            else
            {
                state.mWheelRadius = db.inputs.wheelRadius();
                state.mWheelDistance = db.inputs.wheelDistance();
            }


            if (std::fabs(db.inputs.maxLinearSpeed()) > 0)
            {
                state.mMaxLinearSpeed = std::fabs(db.inputs.maxLinearSpeed());
            }

            if (std::fabs(db.inputs.maxAngularSpeed()) > 0)
            {
                state.mMaxAngularSpeed = std::fabs(db.inputs.maxAngularSpeed());
            }

            if (std::fabs(db.inputs.maxWheelSpeed()) > 0)
            {
                state.mMaxWheelSpeed = std::fabs(db.inputs.maxWheelSpeed());
            }

            state.mInitialized = true;
        }

        double linearVelocity =
            std::max(-state.mMaxLinearSpeed, std::min(state.mMaxLinearSpeed, db.inputs.linearVelocity()));
        double angularVelocity =
            std::max(-state.mMaxAngularSpeed, std::min(state.mMaxAngularSpeed, db.inputs.angularVelocity()));

        // calculate wheel speed
        auto& jointVelocities = db.outputs.velocityCommand();
        jointVelocities.resize(2);
        jointVelocities[0] = ((2 * linearVelocity) - (angularVelocity * state.mWheelDistance)) / (2 * state.mWheelRadius);
        jointVelocities[1] = ((2 * linearVelocity) + (angularVelocity * state.mWheelDistance)) / (2 * state.mWheelRadius);

        jointVelocities[0] = std::max(-state.mMaxWheelSpeed, std::min(state.mMaxWheelSpeed, jointVelocities[0]));
        jointVelocities[1] = std::max(-state.mMaxWheelSpeed, std::min(state.mMaxWheelSpeed, jointVelocities[1]));

        return true;
    }

    virtual void reset()
    {
        GraphObj graphObj{ nodeObj.iNode->getGraph(nodeObj) };
        GraphContextObj context{ graphObj.iGraph->getDefaultGraphContext(graphObj) };

        // set the node's input and output
        AttributeObj linearAttr = nodeObj.iNode->getAttribute(nodeObj, "inputs:linearVelocity");
        auto linearHandle = linearAttr.iAttribute->getAttributeDataHandle(linearAttr, kAccordingToContextIndex);
        double* linearVelocity = getDataW<double>(context, linearHandle);
        *linearVelocity = 0;

        // set the node's input and output
        AttributeObj angularAttr = nodeObj.iNode->getAttribute(nodeObj, "inputs:angularVelocity");
        auto angularHandle = angularAttr.iAttribute->getAttributeDataHandle(angularAttr, kAccordingToContextIndex);
        double* angularVelocity = getDataW<double>(context, angularHandle);
        *angularVelocity = 0;

        // set the node's input and output
        // TODO: Disabled because this causes a crash on reset when next compute is called
        // AttributeObj velocityAttr = nodeObj.iNode->getAttribute(nodeObj, "outputs:velocityCommand");
        // auto velocityHandle = velocityAttr.iAttribute->getAttributeDataHandle(velocityAttr,
        // kAccordingToContextIndex); double* velocityCommand = getDataW<double>(context, velocityHandle);
        // velocityCommand[0] = 0.0;
        // velocityCommand[1] = 0.0;
    }

private:
    bool mInitialized = false;
    double mMaxAngularSpeed = 1.0e7;
    double mMaxWheelSpeed = 1.0e7;
    double mMaxLinearSpeed = 1.0e7;
    double mWheelDistance = 0;
    double mWheelRadius = 0;
    NodeObj nodeObj;
};

REGISTER_OGN_NODE()
}
}
}
